velox/exec/Window.h

```diff
@@ -0,0 +1,228 @@
+      const std::shared_ptr<const core::WindowNode>& windowNode);
+
+  void addInput(RowVectorPtr input) override;
+
+  RowVectorPtr getOutput() override;
+
+  bool needsInput() const override {
+    return !noMoreInput_;
+  }
+
+  void noMoreInput() override;
+
+  BlockingReason isBlocked(ContinueFuture* /* unused */) override {
+    return BlockingReason::kNotBlocked;
+  }
+
+  bool isFinished() override {
+    return finished_;
+  }
+
+ private:
+  // Structure for the window frame for each function.
+  struct WindowFrame {
+    const core::WindowNode::WindowType type;
+    const core::WindowNode::BoundType startType;
+    const core::WindowNode::BoundType endType;
+    const std::optional<column_index_t> startChannel;
+    const std::optional<column_index_t> endChannel;
+  };
+
+  // Helper function to create WindowFunction and frame objects
+  // for this operator.
+  void createWindowFunctions(
+      const std::shared_ptr<const core::WindowNode>& windowNode,
+      const RowTypePtr& inputType);
+
+  // Helper function to create the buffers for peer and frame
+  // row indices to send in window function apply invocations.
+  void createPeerAndFrameBuffers();
+
+  // Function to compute the partitionStartRows_ structure.
+  // partitionStartRows_ is vector of the starting rows index
+  // of each partition in the data. This is an auxiliary
+  // structure that helps simplify the window function computations.
+  void computePartitionStartRows();
+
+  // This function is invoked after receiving all the input data.
+  // The input data needs to be separated into partitions and
+  // ordered within it.
+  void sortPartitions();
+
+  // Helper function to call WindowFunction::resetPartition() for
+  // all WindowFunctions.
+  void callResetPartition(vector_size_t idx);
+
+  // Helper method to call WindowFunction::apply to all the rows
+  // of a partition between startRow and endRow. The outputs
+  // will be written to the vectors in windowFunctionOutputs
+  // starting at offset resultIndex.
+  void callApplyForPartitionRows(
+      vector_size_t startRow,
+      vector_size_t endRow,
+      const std::vector<VectorPtr>& windowFunctionOutputs,
+      vector_size_t resultIndex);
+
+  // Helper function to find the frame end-points for the current
+  // row.
+  std::pair<vector_size_t, vector_size_t> findFrameEndPoints(
+      vector_size_t idx,
+      vector_size_t partitionStartRow,
+      vector_size_t partitionEndRow,
+      vector_size_t currentRow);
+
+  // Helper function to compare the rows at lhs and rhs pointers
+  // using the keyInfo in keys. This can be used to compare the
+  // rows for partitionKeys, orderByKeys or a combination of both.
+  inline bool compareRowsWithKeys(
+      const char* lhs,
+      const char* rhs,
+      const std::vector<std::pair<column_index_t, core::SortOrder>>& keys);
+
+  // Function to compute the window function output values for the
+  // current rows being output.
+  void computeWindowOutputs(
```

Comment:
This function appears to be similar to callApplyForPartitionRows, but their names are quite different. Would it make sense to unify?

Reply:
I changed this to callApplyLoop for more consistency.